### DISCUSSION TITLE
How to create pages from a JSON or excel sheet at build time or dynamically

### DISCUSSION THREAD

**Discussion by CaptainT33mo** [vercel/next.js#11936] [Help]
Hi,
I was previously using Gatsby for our static site generation but we moved to Next JS recently to create dynamic pages with SSR for better SEO. In our Gatsby site, we had 20-30 pages created using JSON data and around 5.5K pages created using Excel Sheet data which was statically generated during the build time. Since we completely moved to Next JS so we want to create those pages in Next JS via JSON or Excel Sheet. So how to create that. I googled it and couldn't find any proper solution to this hence asking it here.

Any help would be much-appreciated guys.

**Comment by janus-reith** [comment]
You would probably use `getStaticPaths` together with `getStaticProps` on some page with dynamic routing.
In `getStaticPaths` you define the list of different url slugs to be created for your pages. Then `getStaticProps` recieves that slug, in there you get the necessary data from your excel files.
You could use excel parser module both in `getStaticPaths` and `getStaticProps`, and it won't be included in your browser bundle.

This eplains it well:
https://nextjs.org/docs/basic-features/data-fetching#getstaticpaths-static-generation

**Reply by CaptainT33mo** [reply]
Hi, @janus-reith Thanks for the reply. I'll definitely try this out.

Also, I have one more question, will these pages be listed in the sitemap of the site?

**Reply by janus-reith** [reply]
Depends on how you create the sitemap, but there is no sitemap out of the box and AFAIK no plug and play solution for that yet.
You could create an api route for that purpose, so you have something like `/pages/api/sitemap`.
In there, you could use a similar logic as in `getStaticPaths` to determine your routes from your excel files and return the xml. Depending on the complexity of that operation you might want to take measures to cache the output somehow, so you don't have to do that for each new request.
Then you create a rewrite in your `next.config.js` so /sitemap.xml is rewritten to /pages/api/sitemap.xml.

While this requires some effort, it gives you very fine-grained control over how your Sitemap should work.
